src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+					// fine: direct callers get to know that we're not actually blocking,
+					// and indirect callers have to assume the worse.
+					// Still output a diagnostic so that we can suggest a code fix in the
+					// IDE to add [Blocking].
+
+					// TODO: code fix to add [Blocking]
+
+					var decl = methodSymbol.DeclaringSyntaxReferences.First().GetSyntax() as MethodDeclarationSyntax;
+
+					ctx.ReportDiagnostic(
+						Diagnostic.Create(
+							Diagnostics.NonBlockingImplementationOfBlockingThing,
+							decl.Identifier.GetLocation(),
+							$"{methodSymbol.ContainingType.Name}.{methodSymbol.Name}",
+							$"{implementedBlockingThings[0].ContainingType.Name}.{implementedBlockingThings[0].Name}"
+						)
+					);
+					return;
+				}
+			}
+
+			// So we do have a [Blocking] attribute.
+
+			if( implementedBlockingThings.Count == 0 ) {
+				// We're not forced to have [Blocking] due to what we implement, so check
+				// if we need it at all.
+				noteMethodWithPossiblyUnusedBlockingAttribute( methodSymbol );
+			}
+
+			if( methodSymbol.IsAsync ) {
+				// TODO: code fix to remove [Blocking]
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotBeBlocking,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						methodSymbol.Name
+					)
+				);
+				return;
+			}
+
+			// Don't let us use [Blocking] if we're implementing something non-blocking,
+			// callers need to know that they may block.
+			if( implementedNonBlockingThings.Count > 0 ) {
+				// TODO: code fix to remove [Blocking]
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.DontIntroduceBlockingInImplementation,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						$"{methodSymbol.ContainingType.Name}.{methodSymbol.Name}",
+						// Just use the first one as an example:
+						$"{implementedNonBlockingThings[0].ContainingType.Name}.{implementedNonBlockingThings[0].Name}"
+					)
+				);
+				return;
+			}
+
+			Debug.Assert(
+				// We are [Blocking]
+				attrData != null
+				// everything we implemement is too
+				&& implementedNonBlockingThings.Count == 0
+			);
+		}
+
+		private static void InspectMethodCall(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodHasBlockingCall
+		) {
+			// Goal: find calls to blocking methods inside non-blocking methods, or in properties.
+
+			var invocation = (InvocationExpressionSyntax)ctx.Node;
+
+			var invokedSymbolInfo = ctx.SemanticModel.GetSymbolInfo( invocation.Expression );
+
+			// Unfortunately there are ways to use this to dodge analysis, but
+			// there is only so much we can do.
+			if( invokedSymbolInfo.Symbol?.Kind != SymbolKind.Method ) {
+				return;
+			}
+
+			var invokedMethodSymbol = (IMethodSymbol)invokedSymbolInfo.Symbol;
+
+			// async methods are never [Blocking] (see other parts of the analyzer)
+			if( invokedMethodSymbol.IsAsync ) {
+				return;
+			}

```

Comment:
Is this visible cross-assembly? Just an optimization to skip the attribute check, regardless, yes?

Reply:
It's not. Ya when I originally wrote it it was an optimization but like the code has changed a bunch since then and I never gave it much thought -- I'll remove it because there is already enough going on in here.

Reply:
Oh, also early attempts to avoid spurious diagnostics, like if you were making a function be async then don't bug all of the callers about how they need to be blocking/start bugging them that they can remove blocking, but switching from sync -> async requires a bunch of changes at the call site anyway so I don't think this is legit.